### PR TITLE
Improve sqlite3 types

### PR DESCRIPTION
**Pull request by JelleZijlstra**

### CODE REVIEW

**Comment by JelleZijlstra** [comment]
The new errors are correct but possibly too pedantic. I made `fetchone()` return `Any | None`, because it can in fact return None: https://github.com/python/cpython/blob/055760ed9e745a3104acbfa8a3b76eb26a72590d/Modules/_sqlite/cursor.c#L829. But often you know that there's at least one row (e.g., because you ran a COUNT query), so I'd be open to changing it back to Any. This is similar to the issues we've had with `re.match` possibly returning None.

**Comment by AlexWaygood** [comment]
> The new errors are correct but possibly too pedantic. I made `fetchone()` return `Any | None`, because it can in fact return None: https://github.com/python/cpython/blob/055760ed9e745a3104acbfa8a3b76eb26a72590d/Modules/_sqlite/cursor.c#L829. But often you know that there's at least one row (e.g., because you ran a COUNT query), so I'd be open to changing it back to Any. This is similar to the issues we've had with `re.match` possibly returning None.

If the aim is to get these new stubs in before the next mypy release, maybe it'd be best to change it back to `Any` for now, and we can revisit it later?

**Review by srittau — CHANGES_REQUESTED** [review]

File: stdlib/sqlite3/dbapi2.pyi [hunk 1]
```diff
@@ -148,22 +155,44 @@ class _WindowAggregateClass(Protocol):
...
-    InterfaceError: Any
-    InternalError: Any
-    NotSupportedError: Any
-    OperationalError: Any
-    ProgrammingError: Any
-    Warning: Any
-    in_transaction: Any
-    isolation_level: Any
+    @property
+    def DataError(self) -> type[sqlite3.DataError]: ...
```
[the remark this answers is not in the document]
Comment (JelleZijlstra) [reply stdlib/sqlite3/dbapi2.pyi:159]: Yes I got a little creative here. They're read-only properties at runtime; I have no idea why they're exposed to Python code. Importing `sqlite3` helps avoid a name conflict here.

File: stdlib/sqlite3/dbapi2.pyi [hunk 2]
```diff
@@ -187,17 +216,21 @@ class Connection:
             self, __name: str, __num_params: int, __aggregate_class: Callable[[], _WindowAggregateClass] | None
         ) -> None: ...
 
-    def create_collation(self, __name: str, __callback: Any) -> None: ...
+    def create_collation(self, __name: str, __callback: Callable[[str, str], None] | None) -> None: ...
```
[the remark this answers is not in the document]
Comment (JelleZijlstra) [reply stdlib/sqlite3/dbapi2.pyi:219]:
```suggestion
    def create_collation(self, __name: str, __callback: Callable[[str, str], int | SupportsIndex] | None) -> None: ...
```
It actually accepts any int; I know the docs say only -1, 0, or 1.

**Review by srittau — APPROVED** [review]

**Pull request merged**